tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,681 @@
+            if curr & CLOSED > 0 {
+                return Ready(Err(AcquireError::closed()));
+            }
+
+            let mut remaining = 0;
+            let (next, acq) = if curr + acquired >= needed {
+                let next = curr - (needed - acquired);
+                (next, needed)
+            } else {
+                remaining = (needed - acquired) - curr;
+                (0, curr)
+            };
+
+            if remaining > 0 && lock.is_none() {
+                // No permits were immediately available, so this permit will
+                // (probably) need to wait. We'll need to acquire a lock on the
+                // wait queue.
+                if let Ok(l) = self.waiters.try_lock() {
+                    // If we were able to acquire the lock *now* without
+                    // blocking, we don't need to reload the semaphore's permit state.
+                    lock = Some(l)
+                } else {
+                    lock = Some(self.waiters.lock().unwrap());
+                    // While we were waiting to lock the wait list, additional
+                    // permits may have been released. Therefore, we will acquire a
+                    // new snapshot of the current state of the semaphore before
+                    // actually enqueueing the waiter..
+                    curr = self.permits.load(Acquire);
+                    continue;
+                }
+            }
+
+            match self
+                .permits
+                .compare_exchange_weak(curr, next, AcqRel, Acquire)
+            {
+                Ok(_) => {
+                    acquired += acq;
+                    if remaining == 0 && !queued {
+                        return Ready(Ok(()));
+                    }
+                    break lock.unwrap();
+                }
+                Err(actual) => curr = actual,
+            }
+            drop(lock.take());
+        };
+
+        if node.assign_permits(&mut acquired) {
+            if acquired > 0 {
+                // We ended up with more permits than we needed. Release the
+                // back to the semaphore.
+                self.add_permits_locked(acquired, waiters);
+            }
+            return Ready(Ok(()));
+        }
+
+        assert_eq!(acquired, 0);
+
+        // Otherwise, register the waker & enqueue the node.
+        unsafe {
+            node.waker.with_mut(|waker| {
+                // Safety: the wait list is locked, so we may modify the waker.
+                *waker = Some(cx.waker().clone());
+            });
+
+            let node = Pin::into_inner_unchecked(node) as *mut _;
+            let node = NonNull::new_unchecked(node);
+
+            // If the waiter is not already in the wait queue, enqueue it.
+            if !waiters.contains(node) {
+                waiters.queue.push_front(node);
+            }
+        }
+
+        Pending
+    }
+}
+
+impl Drop for Semaphore {
+    fn drop(&mut self) {
+        self.close();
```

Comment:
Is this needed? In theory, there can be no waiters at this point.

Reply:
Hmm, in that case, I believe the `close` method _could_ be removed entirely, since this is the only place it's called outside of tests. However, if we eventually intend to port the bounded MPSC to use the new semaphore implementation, we _will_ need a `close` method (and all the associated state for managing closing)...do you think it's better to leave the `close` method with an `#[allow(dead_code)]` given that, or should it be removed for now?